hail/python/hail/expr/functions.py

```diff
@@ -128,6 +128,50 @@ def compute_global_error(s):
         return hl.rbind(cdf, lambda cdf: hl.rbind(compute_sum(cdf), compute_single_error))
 
 
+def _error_from_cdf_python(cdf, failure_prob, all_quantiles=False):
+    """Estimates error of approx_cdf aggregator, using Hoeffding's inequality.
+
+    Parameters
+    ----------
+    cdf : :obj:`dict`
+        Result of :func:`.approx_cdf` aggregator, evaluated to a python dict
+    failure_prob: :obj:`float`
+        Upper bound on probability of true error being greater than estimated error.
+    all_quantiles: :obj:`bool`
+        If ``True``, with probability 1 - `failure_prob`, error estimate applies
+        to all quantiles simultaneously.
+
+    Returns
+    -------
+    :obj:`float`
+        Upper bound on error of quantile estimates.
+    """
+    import math
+
+    s = 0
+    for i in builtins.range(builtins.len(cdf._compaction_counts)):
+        s += cdf._compaction_counts[i] << (2 * i)
+    s = s / (cdf.ranks[-1] ** 2)
+
+    def update_grid_size(p):
+        return 4 * math.sqrt(math.log(2 * p / failure_prob) / (2 * s))
+
+    def compute_grid_size(s):
+        p = 1 / failure_prob
+        for _ in builtins.range(5):
+            p = update_grid_size(p)
+        return p
+
+    def compute_single_error(s, failure_prob=failure_prob):
```

Comment:
Do you intend to pull these inner functions out or can we just get `failure_prob` from the enclosing scope (ie remove the parameter)?

Reply:
The parameter is there because it gets called with different values. This function computes the error bound on an estimated rank for a single value, with a given probability of exceeding the bound. We sometimes want an error bound that applies to *all* possible values simultaneously, with a given probability of *any* rank estimate exceeding the bound. Computing that involves computing an error bound for a single value with an appropriately smaller failure probability.